moveit_core/collision_detection_fcl/src/collision_common.cpp

```diff
@@ -586,12 +592,17 @@ bool distanceCallback(fcl::CollisionObjectd* o1, fcl::CollisionObjectd* o2, void
 #if (MOVEIT_FCL_VERSION >= FCL_VERSION_CHECK(0, 6, 0))
         dist_result.nearest_points[0] = contact.pos;
         dist_result.nearest_points[1] = contact.pos;
-        dist_result.normal = contact.normal;
+        Eigen::Vector3d normal = contact.normal;
 #else
         dist_result.nearest_points[0] = Eigen::Vector3d(contact.pos.data.vs);
         dist_result.nearest_points[1] = Eigen::Vector3d(contact.pos.data.vs);
-        dist_result.normal = Eigen::Vector3d(contact.normal.data.vs);
+        Eigen::Vector3d normal(contact.normal.data.vs);
```

Comment:
It's unfortunate that there will be a small hit to performance here. We have an application where fast distance checks are important.

I guess it is worthwhile, though. Any other thoughts on the impact on performance?

Reply:
This branch of code is only executed when `enable_signed_distance` is true, so the performance impact will only be visible for these cases.

I think we should definitely fix it, since the current code will report a normal in the wrong direction.

Reply:
I would not expect a performance drop at all (assuming compiler optimizations are enabled).
The allocated variable is only a helper because of the `#ifdef` and should be optimized away.

Reply:
I'm always skeptical when somebody says, "just trust compiler optimizations."

Attached is some data from a BioIK control loop that does probably thousands of distance checks per second. On main branch, the longest duration is < `0.004s`. With this PR, the longest control period is` 0.0112s`.

I would like to merge the PR too but I don't like this.

[signed_distance_pr.txt](https://github.com/ros-planning/moveit/files/6539790/signed_distance_pr.txt)
[main.txt](https://github.com/ros-planning/moveit/files/6539791/main.txt)

Reply:
Hm, that's a pretty significant increase... I'd guess the issue is not this particular line (it seems rather innocent and optimizable to me as well), but either the branch below (which inverts the normal) or the `res_cd1`/`res_cd2` stuff above, which definitely causes a memory read. But still, I'm quite surprised. The `distanceCallback` is called once per collision object pair, correct? My feeling was that anything we do here should not have a huge impact compared to the stuff FCL has to do internally on each `fcl::distance()` call.

Are you using just primitives or meshes as well?

Do you require the `nearest_points` / `normal` results in your application? Right now `enable_signed_distance` implies `nearest_points` and `normal` calculation, but I assume you don't actually use the normal (since it's wrong half of the time). Maybe an acceptable solution for your use case would be to add an additional `if (cdata->req->enable_nearest_points)`  around normal computation even in the signed distance case?

I could try to set up a loop that repeatedly evaluates distances and to get some insight using a profiler.

Reply:
We're using meshes. Yeah, we don't care about the normal, just care about the distance.

I like this idea -- 

` Maybe an acceptable solution for your use case would be to add an additional if (cdata->req->enable_nearest_points) around normal computation even in the signed distance case?`

If you want to push that, I can test it easily.

Reply:
> I'm always skeptical when somebody says, "just trust compiler optimizations."

You better be! That's what benchmarks are for and it's why I used "expect".
Though you are the one who said he's not interested in GSoC projects for better Benchmarking. :-)
Thanks for updating the name by the way :medal_sports: 

I'm quite curious where this delay comes from, for educational reasons. But I also think the additional delay should not be that severe.

On optimizing the runtime: I might not be deep enough in this code at the moment, but wouldn't you just need one additional comparison `o1 == fcl_result.o1` to distinguish between the original case and the failure condition you found (inverted objects). We can't get around the negation and the pointer comparison, but that should be everything in the ideal case.

That being said, additionally guarding the lines that fill in nearest points/normal when using `enable_signed_distance` with `enable_nearest_points` makes some sense too, but it breaks method semantics in a bad way. :crying_cat_face: 
In case that's actually required to keep the rough current performance please update the logic to compute the distance only once (it's already computed with worth accuracy above) and add an entry to the [`MIGRATION.md`](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md)

Reply:
It looks like the average control loop period is within 10% now so I'd be happy merging this as-is.  :+1:

Master:  mean period = 0.0045s, worst case = 0.014s
This PR: mean period = 0.0052s, worst case = 0.013s

(Maybe I just screwed up the data collection before?)